prow/test/integration/test/gangway_test.go

```diff
@@ -0,0 +1,325 @@
+	"testing"
+	"time"
+
+	"github.com/google/go-cmp/cmp"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+	"google.golang.org/grpc/metadata"
+	coreapi "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/kubernetes"
+	ctrlruntimeclient "sigs.k8s.io/controller-runtime/pkg/client"
+
+	prowjobv1 "k8s.io/test-infra/prow/apis/prowjobs/v1"
+	"k8s.io/test-infra/prow/gangway"
+	"k8s.io/test-infra/prow/kube"
+	"k8s.io/test-infra/prow/test/integration/internal/fakegitserver"
+)
+
+// TestGangway makes gRPC calls to gangway.
+func TestGangway(t *testing.T) {
+	t.Parallel()
+
+	const (
+		UidLabel         = "integration-test/uid"
+		ProwJobDecorated = `
+presubmits:
+  - name: trigger-inrepoconfig-presubmit-via-gangway%s
+    always_run: false
+    decorate: true
+    spec:
+      containers:
+      - image: localhost:5001/alpine
+        command:
+        - sh
+        args:
+        - -c
+        - |
+          set -eu
+          echo "hello from trigger-inrepoconfig-presubmit-via-gangway-repo%s"
+          cat README.txt
+`
+	)
+
+	// createGerritRepo creates Gerrit-style Git refs for changes (PRs). The
+	// revision is always named "refs/changes/00/123/1" where 123 is the change
+	// number (PR number) and 1 is the first revision (version) of the same
+	// change.
+	CreateRepo1 := createGerritRepo("TestGangway1", fmt.Sprintf(ProwJobDecorated, "1", "1"))
+
+	tests := []struct {
+		name       string
+		repoSetups []fakegitserver.RepoSetup
+		metadata   []string
+		msg        *gangway.CreateJobExecutionRequest
+		want       string
+	}{
+		{
+			name: "inrepoconfig-presubmit",
+			repoSetups: []fakegitserver.RepoSetup{
+				{
+					Name:      "some/org/gangway-test-repo-1",
+					Script:    CreateRepo1,
+					Overwrite: true,
+				},
+			},
+			metadata: []string{
+				"x-endpoint-api-consumer-type", "PROJECT",
+				"x-endpoint-api-consumer-number", "123"},
+			msg: &gangway.CreateJobExecutionRequest{
+				JobDefinition: &gangway.JobDefinition{
+					Name:          "trigger-inrepoconfig-presubmit-via-gangway1",
+					ExecutionType: gangway.JobExecutionType_PRESUBMIT,
+					// Define where the job definition lives from inrepoconfig.
+					InrepoconfigSource: &gangway.GitReferenceDynamic{
+						Base: &gangway.GitReferenceStatic{
+							Url:    "https://fakegitserver.default/repo/some/org/gangway-test-repo-1",
+							Commit: "f1267354a7bbc5ce7d0458cdf4d0d36e8d35d8b3",
+							Ref:    "master",
+						},
+						RefsToMerge: []*gangway.GitReferenceStatic{
+							{
+								Url:    "https://fakegitserver.default/repo/some/org/gangway-test-repo-1",
+								Commit: "458b96a96a74689447530035f5a71c426bacb505",
+								Ref:    "refs/changes/00/123/1",
+							},
+						},
+					},
+				},
+				// Define which refs will be cloned and merged before running
+				// the job. This is traditionally the same as the
+				// InrepoconfigSource but there could be use cases where this
+				// 1:1 mapping is not desirable.
```

Comment:
@cjwagner FYI

Reply:
The 1:1 mapping is not enforced with the current design. That is, you could construct a CreateJobExecutionRequest message without a BaseRepo field (and always make it so that this BaseRepo field matches either the InrepoconfigSource or the orgRepo defined statically in the job config). But I chose to do it this other way because I felt that it gave more flexibility.

Reply:
I don't understand how it would make sense for this to differ from the inrepoconfigsource. Can you share a concrete example of the flexibility you are talking about?

Reply:
At a high level this goes back to the "how jobs are configured" vs "how jobs are run" idea (job definitions vs executions).

In Prow today we have implicit refs. If you define a presubmit job under `myOrg/myRepo`, `myRepo` is cloned automatically. Likewise, if you define a job with inrepoconfig, that repo is cloned automatically.

One example is the idea of reusable job configurations. If we break away from implicit refs entirely (as is done in this PR), then we can let users run "canonical" jobs like "go test" "make lint" or some such without forcing them to define the job themselves for their `myRepo`.

Reply:
I also want to add that I want the API to be as flexible as possible, because it creates a brand new Prow Job CR so it has every opportunity to allow clients to shape the final CR. I am more concerned about unknowingly forcing our users into a corner.

One additional idea I had was that we could add some basic helper functions (for clients) so that they don't have to see this flexibility. Like we could have a function that assumes the old mindset of implicit refs so that if they use that function, the generated CreateJobExecutionRequest does a 1:1 copy of the Git refs from the JobDefinition into the BaseRepo.

Reply:
Just a quick note to capture the current thought here: we discussed more offline and decided that having a 1:1 mapping is the way to go for now. If we need the separation in the future, we can add that functionality (it would basically involve adding some fields and associated business logic). IOW, we can remove the BaseRepo field for now and can consider adding it back later. In the interim, the Git information from the JobDefinition could be "promoted" to reside in the CreateJobExecutionRequest message directly, just like how we expect a "refs" payload for triggering via Sub (for presubmits and postsubmits). That is, the Git information will no longer be tied solely to the JobDefinition and will be responsible for both fetching job information and/or encoding information for which Git repo to clone in the test environment, or both (as is the case for triggering via Sub). See https://docs.prow.k8s.io/docs/components/optional/sub/.